### PR TITLE
build config: Gradle Version Catalog + type-safe project dependencies

### PR DESCRIPTION
**Pull request by aSemy**
* Moves versions out of `gradle.properties` and into `libs.versions.toml` [version catalog](https://docs.gradle.org/current/userguide/platforms.html)
* Uses [type-safe project dependencies](https://docs.gradle.org/current/userguide/declaring_dependencies.html#sec:type-safe-project-accessors) for internal projects
* refactor the maven-cli downloader convention plugin (added in #2911) to use the version catalog, and for clarity
* Additionally updates JUnit to 5.9.2

Depends on #2652

### CODE REVIEW

File: gradle/libs.versions.toml
Comment (Goooler) [inline gradle/libs.versions.toml:13]: We can inline these once-use versions.

File: integration-tests/build.gradle.kts
Comment (Goooler) [inline integration-tests/build.gradle.kts:10]:
```suggestion
```

File: kotlin-analysis/intellij-dependency/build.gradle.kts
Comment (Goooler) [inline kotlin-analysis/intellij-dependency/build.gradle.kts:31]: Why not put these into toml?

File: plugins/base/base-test-utils/build.gradle.kts
Comment (Goooler) [inline plugins/base/base-test-utils/build.gradle.kts:16]:
```suggestion
```

**Review by Goooler — COMMENTED** [review]
Good to see! We can migrate to type-safe project accessors in next PR.

**Review by IgnatBeresnev — COMMENTED** [review]
Had a glance - looking good, thanks!

I'll have a closer look once #2704 gets merged and this PR is rebased onto it

File: kotlin-analysis/intellij-dependency/build.gradle.kts
Comment (aSemy) [reply kotlin-analysis/intellij-dependency/build.gradle.kts:31]: this PR is still WIP (but I forgot to mark it). I got started with the Version Catalog in #2704, but decided to split it out and I made this PR so I could put a link next to the TODO https://github.com/Kotlin/dokka/pull/2704/files#diff-7b5c40eb5522b5216cfe8c75612e579a8db0daee9ded80df9ce6a56e8361a962R11

**Comment by aSemy** [comment]
> Since there are a bunch of conflicts, proposing to merge #2652 first, it's been waiting for a while :) Hopefully shouldn't take as long

Agreed, sounds good!

**Comment by IgnatBeresnev** [comment]
The PR is ready to be rebased :) Looking forward to it!

Please, if you can, try to address copy-pasted versions that I mentioned in https://github.com/Kotlin/dokka/pull/2912#discussion_r1133791517, I assumed that we'd have the actual version strings in the toml only

File: build-logic/src/main/kotlin/org/jetbrains/publication.kt
Comment (aSemy) [inline build-logic/src/main/kotlin/org/jetbrains/publication.kt:149]: this is just some unrelated tidying & formatting

File: examples/gradle/dokka-customFormat-example/build.gradle.kts
Comment (aSemy) [inline examples/gradle/dokka-customFormat-example/build.gradle.kts:34]: > You no longer need to declare a dependency on the `stdlib` library in any Kotlin Gradle project, including a multiplatform one. The dependency is added by default.

https://kotlinlang.org/docs/whatsnew14.html#dependency-on-the-standard-library-added-by-default
 
Therefore I removed it from all subprojects and examples.

File: settings.gradle.kts
Comment (aSemy) [inline settings.gradle.kts:6]: I moved `includeBuild` into the plugins block [as recommended](https://docs.gradle.org/current/userguide/composite_builds.html#included_plugin_builds)

> You may also use the `includeBuild` mechanism outside `pluginManagement` to include plugin builds. However, this does not support all use cases and including plugin builds like that might be deprecated in a future Gradle version.

I also moved the repository definitions to be above plugins block, since logically the repos should be defined before plugins are resolved - but that's not important and I can revert if you want.

File: gradle/libs.versions.toml
Comment (aSemy) [reply gradle/libs.versions.toml:7]: I think it's best to define all the versions at the top, because it's consistent, it's easier to scan the top to see the versions, and if any versions do need to be re-used then then it's less work (personally I find editing `libs.versions.toml` a bit fiddly, so I try to avoid it).

File: settings.gradle.kts
Comment (Goooler) [reply settings.gradle.kts:6]: I moved it out cause I noticed https://github.com/detekt/detekt/commit/0245b0f24b032649908efa37f78b6abd2374efee.

File: gradle/libs.versions.toml
Comment (Goooler) [reply gradle/libs.versions.toml:7]: We can easy to know which versions are reused after inlining unnecessary ones, either way.

**Review by Goooler — COMMENTED** [review]
Good to see this, both version catalog and type-safe project accessors!
We can rename this PR's title to something like "Migrate to type-safe dependency accessors"?

File: runners/gradle-plugin/build.gradle.kts
Comment (IgnatBeresnev) [inline runners/gradle-plugin/build.gradle.kts:11]: Previously it was declared as `api(..)` - does it make any difference in the case of a Gradle plugin?

File: kotlin-analysis/intellij-dependency/build.gradle.kts
Comment (IgnatBeresnev) [inline kotlin-analysis/intellij-dependency/build.gradle.kts:48]: Qodana is complaining about unstable API usage

**Review by IgnatBeresnev — APPROVED** [review]
Great to see, as always!

[The main integration tests](https://github.com/Kotlin/dokka/actions/runs/4419616345/jobs/7775239862) are failing, could you please fix it? I'll run longer integration tests after that

File: settings.gradle.kts
Comment (IgnatBeresnev) [reply settings.gradle.kts:6]: Hm. We don't use configuration cache in this project, do we? So we don't have the warning yet, but it might resurface one day

I don't mind leaving it as is, we can deal with the warning when we get it, but if you do move it out - please add a comment why

File: gradle/libs.versions.toml
Comment (IgnatBeresnev) [reply gradle/libs.versions.toml:7]: I think it should be consistent - either all versions are at the top, or all of the versions are inlined, otherwise it might eventually lead to chaos. At the moment, I personally don't have any preference, so I'm fine with either approach :+1:

File: settings.gradle.kts
Comment (Goooler) [reply settings.gradle.kts:6]: Make sense.

File: kotlin-analysis/intellij-dependency/build.gradle.kts
Comment (aSemy) [reply kotlin-analysis/intellij-dependency/build.gradle.kts:48]: I'll suppress it for now, but tbh I think this bit of code probably needs a refactor - the JARs can be fetched via existing configurations (possibly). 

I think these dependencies are related to https://github.com/Kotlin/dokka/issues/2889 too? So might they be removed in the future?

File: settings.gradle.kts
Comment (aSemy) [reply settings.gradle.kts:6]: I don't think the warning in https://github.com/detekt/detekt/pull/5690 is about configuration cache, it's about the (configuration of) build cache.

https://github.com/detekt/detekt/blob/b07acae2b5ce2d5c8e8216a4e3e022f89c5fa67f/settings.gradle.kts#L71-L84

But Dokka doesn't have that (JetBrains really should set up a Gradle Build Cache server though!)

File: runners/gradle-plugin/build.gradle.kts
Comment (aSemy) [reply runners/gradle-plugin/build.gradle.kts]: Good spot, thanks

File: runners/gradle-plugin/build.gradle.kts
Comment (aSemy) [inline runners/gradle-plugin/build.gradle.kts:15]: I removed `compileOnly(gradleKotlinDsl())` because it's added by the `kotlin-dsl` plugin

File: integration-tests/gradle/projects/serialization/kotlinx-serialization
Comment (aSemy) [inline integration-tests/gradle/projects/serialization/kotlinx-serialization:1]: I think I accidentally updated the git commit of KxS in my branch. How can I revert it back to the commit that dokka/master uses...?

File: kotlin-analysis/intellij-dependency/build.gradle.kts
Comment (IgnatBeresnev) [reply kotlin-analysis/intellij-dependency/build.gradle.kts:50]: Yep, suppressing it is totally fine. These should be removed/refactored quite soon, it's a blocked for 1.8.20

File: integration-tests/gradle/projects/serialization/kotlinx-serialization
Comment (IgnatBeresnev) [reply integration-tests/gradle/projects/serialization/kotlinx-serialization:1]: Looks like you were able to figure it out :)

I usually cd into thee KxS project dir within Dokka, and git checkout the needed commit id, and then commit the change

Comment (aSemy) [reply integration-tests/gradle/projects/serialization/kotlinx-serialization:1]: might be fixed now? 

For future reference, in terminal:

* `cd` into the KxS directory
* `git checkout daa95c79ffadc0eedbbb4a481a00556b78212e43` (or whatever the hash is, which can be found by clicking the link to the KxS project in this dir https://github.com/Kotlin/dokka/tree/master/integration-tests/gradle/projects/serialization)
* `cd ..` back up to Dokka root dir
* commit the KxS directory

Comment (aSemy) [reply integration-tests/gradle/projects/serialization/kotlinx-serialization:1]: btw, it's possible to use Gradle to download GitHub projects as if they were dependencies. This might be a nice replacement because the version can be set explicitly, but the current approach also works.

https://stackoverflow.com/a/34327202/4161471

File: settings.gradle.kts
Comment (IgnatBeresnev) [reply settings.gradle.kts:6]: > JetBrains really should set up a Gradle Build Cache server though!

JFYI we actually have a JetBrains instance for Gradle Enterprise (ge.jetbrains.com), but it didn't look like it using it would benefit us in any way compared to what we have now. If there's some really cool things that can be done with it that will work with Dokka, we can switch to it, no problem

**Comment by IgnatBeresnev** [comment]
Triggered teamcity integration tests, can be merged once they pass

File: gradle/libs.versions.toml
Comment (IgnatBeresnev) [inline gradle/libs.versions.toml:62]: Doesn't seem to fail the tests (yet), but maybe it's worth using `*-jvm` as before, to avoid possible problems. If a problem with it surfaces, I don't want to have to roll back the whole PR, so imo it'd be nice to keep the artifacts the same as before, and any additional improvements separately

btw, the cli version 0.3.4 could be extracted into a variable, probably forgotten?

```suggestion
kotlinx-cli = { module = "org.jetbrains.kotlinx:kotlinx-cli-jvm", version = "0.3.4" }
kotlinx-html = { module = "org.jetbrains.kotlinx:kotlinx-html-jvm", version.ref = "kotlinx-html" }
```

File: settings.gradle.kts
Comment (aSemy) [reply settings.gradle.kts:6]: If there's public, read-only access to the JetBrains Gradle build cache server then that would definitely help! It means that everyone's builds would be faster.

https://docs.gradle.org/current/userguide/build_cache.html#sec:build_cache_configure_remote

**Pull request merged**
